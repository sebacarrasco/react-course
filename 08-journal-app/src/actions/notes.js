import { db } from "../firebase/firebase-config";
import { types } from "../types/types";
import { loadNotes } from '../helpers/loadNotes';
import Swal from 'sweetalert2';
import { fileUpload } from "../helpers/fileUpload";

// react-journal

export const startNewNote = () => {
    // Así es, aparte de dispatch se puede acceder a getState que
    // retorna un objeto con tooodos los estados en la store.
    // Bastante similar a useSelector (pero no recibe un callback)
    return async(dispatch, getState) => {

        const { uid } = getState().auth;
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }
        try {
            // Dentro de collection va el path donde se quiere grabar
            const documentReference = await db.collection(`${ uid }/journal/notes`).add(newNote);
            dispatch(activeNote(documentReference.id, newNote));
            dispatch(addNewNote(documentReference.id, newNote));
        } catch (error) {
            console.log(error);
        }
    }
}

export const activeNote = (id, note) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note
    }
});

export const addNewNote = (id, note) => ({
    type: types.notesAddNew,
    payload: {
        id,
        ...note
    }
})

export const startLoadingNotes = ( uid ) => {
    return async(dispatch) => {
        const notes = await loadNotes(uid);
        dispatch(setNotes(notes));
    }
}

export const setNotes = (notes) => ({
    type: types.notesLoad,
    payload: [...notes]
});


export const startSaveNote = (note) => {
    return async(dispatch, getState) => {

        const { uid } = getState().auth;

        // Si el url es undefined, entonces no lo mandaremos porque
        // firebase no deja guardar algo undefined (tira error)
        if (!note.url)
        {
            delete note.url;
        }

        // Eliminamos el id porque no lo guardamos dentro del objeto en firestore,
        // porque en firestore se llega al objeto a través del id, pero el objeto
        // en sí no contiene en sus propiedades el id (aunque podría tenerlo)
        const noteToFirestore = { ...note };
        delete noteToFirestore.id;

        await db.doc(`${ uid }/journal/notes/${ note.id }`).update(noteToFirestore);

        dispatch(refreshNote(note.id, note));
        Swal.fire('Saved', note.title, 'success');
    }
}

// Esta acción se encarga de actualizar en la store la note que se actualizó en firebase
export const refreshNote = (id, note) => ({
    type: types.notesUpdated,
    payload: {
        id,
        note
    }
})


export const startUploading = (file) => {
    return async (dispatch, getState) => {

        const { active:activeNote } = getState().notes;

        Swal.fire({
            title: 'Uploading...',
            text: 'Please wait...',
            allowOutsideClick: false,
            showConfirmButton: false,
            willOpen: () => {
                Swal.showLoading();
            }
        });

        const fileUrl = await fileUpload(file);
        activeNote.url = fileUrl;

        dispatch(startSaveNote(activeNote));

        Swal.close();
    }
}


export const startDeleting = (id) => {
    return async(disppatch, getState) => {

        const { uid } = getState().auth;
        await db.doc(`${ uid }/journal/notes/${ id }`).delete();

        disppatch(deleteNote(id));

    }
}

export const deleteNote = (id) => ({
    type: types.notesDelete,
    payload: id
})

export const notesLogout = () => ({
    type: types.notesLogoutCleaning
})