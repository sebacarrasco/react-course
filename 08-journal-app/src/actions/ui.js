import { types } from "../types/types";


export const setRegisterError = (error) => ({
    type: types.uiSetRegisterError,
    payload: error
});

export const removeRegisterError = () => ({
    type: types.uiRemoveRegisterError
});


export const setLoginError = (error) => ({
    type: types.uiSetLoginError,
    payload: error
});

export const removeLoginError = () => ({
    type: types.uiRemoveLoginError
});


export const StartLoading = () =>({
    type: types.uiStartLoading
});

export const FinishLoading = () =>({
    type: types.uiFinishLoading
});