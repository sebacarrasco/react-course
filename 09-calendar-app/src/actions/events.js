import Swal from "sweetalert2";
import { fetchWithToken } from "../helpers/fetch";
import { prepareEvents } from "../helpers/prepareEvents";
import { types } from "../types/types";

export const eventStartAddNew = (event) => {
    return async(dispatch, getState) => {

        try {
            
            const response = await fetchWithToken('events', event, 'POST');
            const body = await response.json();

            const { uid, name } = getState().auth;

            if(body.ok)
            {
                event.id = body.event.id;
                event.user = {
                    _id: uid,
                    name
                }
                dispatch(eventAddNew(event));
            }

        } catch (error) {
            console.log(error);
        }

    }
}

export const eventAddNew = (event) => ({
    type: types.eventAddNew,
    payload: event
});

export const eventSetActive = (event) => ({
    type: types.eventSetActive,
    payload: event
});

export const eventClearActiveEvent = () => ({
    type: types.eventClearActiveEvent
});

export const eventStartUpdate = (event) => {

    return async (dispatch) => {

        try {
            const response = await fetchWithToken(`events/${ event.id }`, event, 'PUT');
            const body = await response.json();
            if (body.ok)
            {
                dispatch(eventUpdated(event));
            }
            else
            {
                Swal.fire("Error", body.msg, "error");
            }
        } catch (error) {
            console.log(error);
        }
    }

}

export const eventUpdated = (event) => ({
    type: types.eventUpdated,
    payload: event
});

export const eventStartDelete = () => {

    return async (dispatch, getState) => {

        const { id } = getState().calendar.activeEvent;

        try {
            const response = await fetchWithToken(`events/${ id }`, {},  'DELETE');
            const body = await response.json();
            if (body.ok)
            {
                dispatch(eventDeleted());
            }
            else
            {
                Swal.fire("Error", body.msg, "error");
            }
        } catch (error) {
            console.log(error);
        }
    }

}

export const eventDeleted = () => ({
    type: types.eventDeleted
});

export const eventStartLoading = () => {
    return async(dispatch) => {

        try {
            const response = await fetchWithToken('events');
            const body = await response.json();
            // Pasamos las fechas de string a date
            const events = prepareEvents(body.events);
            dispatch(eventLoaded(events));
        } catch (error) {
            console.log(error);
        }

    }
}

export const eventLoaded = (events) => ({
    type: types.eventLoaded,
    payload: events
});

export const eventLogout = () => ({
    type: types.eventLogout
});