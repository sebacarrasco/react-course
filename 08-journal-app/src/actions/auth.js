import Swal from 'sweetalert2';
import { types } from "../types/types";
import { firebase, googleAuthProvider } from '../firebase/firebase-config';
import { FinishLoading, StartLoading } from "./ui";
import { notesLogout } from './notes';

export const startLoginEmailPassword = (email, password) => {
    // Cuando una acción retorna un callback (una función)
    // el midleware la ejecutará. Además pasa como parámetro
    // el dispatch gracias a thunk
    return (dispatch) => {

        dispatch(StartLoading());

        // El return es solo para que en testing aparezcan las acciones login y finishLoading como
        // despachadas
        return firebase.auth().signInWithEmailAndPassword(email, password)
            .then(({ user }) => {
                dispatch(login(user.uid, user.displayName));
                dispatch(FinishLoading());
            })
            .catch(error => {
                dispatch(FinishLoading());
                Swal.fire('Error', error.message, 'error');
            });
    }
}

export const startRegisterWithEmailPasswordName = (email, password, name) => {
    return (dispatch) => {
        
        dispatch(StartLoading());

        firebase.auth().createUserWithEmailAndPassword(email, password)
            .then( async({ user }) => {
                await user.updateProfile({displayName: name}); // Hay que establecerle el displayName a diferencia
                                                            // de cuando se hace la autenticación con google (se actualiza en db y el objeto)
                dispatch(login(user.uid, user.displayName));
                dispatch(FinishLoading());
            })
            .catch(error => {
                dispatch(FinishLoading());
                Swal.fire('Error', error.message, 'error');
            });
    }
}

export const startGoogleLogin = () => {
    // Como es una tarea asíncrona, se tiene que retornar una función.
    // Nótese que estas acciones son despachadas también.
    // Al ser despachadas y no retornar un objeto sino que una función,
    // entonces se ejecuta la función retornada y se pueden hacer dispatch dentro
    // de la misma función
    return (dispatch) => {
        firebase.auth().signInWithPopup( googleAuthProvider )
            .then( userCred => {
                dispatch(login(userCred.user.uid, userCred.user.displayName))
            })
    }
    // El displayName lo tendremos siempre que se autentica con una red social, github o algo así
}

export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName
    }
});


export const startlogout = () => {
    return async(dispatch) => {
        await firebase.auth().signOut(); //Solo puede fallar por conexión a internet
        dispatch(logout());
        dispatch(notesLogout());
    }
}

export const logout = () => ({
    type: types.logout
});