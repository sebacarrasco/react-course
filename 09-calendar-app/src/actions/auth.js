import { fetchWithoutToken, fetchWithToken } from "../helpers/fetch"
import { types } from "../types/types";
import Swal from 'sweetalert2'
import { eventLogout } from "./events";


export const startLogin = (email, password) => {
    
    return async(dispatch) => {

        const response = await fetchWithoutToken('auth', { email, password }, 'POST');
        const body = await response.json();

        if (body.ok)
        {
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());
            dispatch(login({
                uid: body.uid,
                name: body.name
            }));
        }
        else
        {
            Swal.fire('Error', body.msg, 'error');
        }
    }

}

export const startRegister = (email, password, name) => {

    return async(dispatch) => {

        const response = await fetchWithoutToken('auth/new', { email, password, name }, 'POST');
        const body = await response.json();

        if (body.ok)
        {
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());
            dispatch(login({
                uid: body.uid,
                name: body.name
            }));
        }
        else
        {
            Swal.fire('Error', body.msg, 'error');
        }

    }

}

export const startChecking = () => {

    return async(dispatch) => {

        const response = await fetchWithToken('auth/renew');
        const body = await response.json();

        if (body.ok)
        {
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());
            dispatch(login({
                uid: body.uid,
                name: body.name
            }));
        }
        else
        {
            dispatch(checkingFinish());
        }

    }

}

export const checkingFinish = () => ({
    type: types.authCheckingFinish
});

export const login = (user) => ({
    type: types.authLogin,
    payload: user
});

export const startLogout = () => {

    return (dispatch) => {
        localStorage.clear();
        dispatch(eventLogout());
        dispatch(logout());
    }

}

export const logout = () => ({
    type: types.authLogout
})