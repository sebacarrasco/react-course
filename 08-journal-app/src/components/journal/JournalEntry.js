import React from 'react';
import moment from 'moment';
import { useDispatch } from 'react-redux';
import { activeNote } from '../../actions/notes';

export const JournalEntry = ({ id, date, title, body, url }) => {

    const noteDate = moment(date);

    const dispatch = useDispatch();

    const handleEntryClick = () => {
        dispatch(activeNote(id, { title, body, date, url }));
    }

    return (
        <div
            className="journal__entry pointer animate__animated animate__fadeIn animate__faster"
            onClick={ handleEntryClick }
        >
            
            {/* Se pone un div en vez de img para jugar con el cover de la imagen por si
            no es cuadrada o algo as√≠ */}
            {
                // Solo si el url no es undefined se muestra la imagen
                url &&
                (<div
                    className="journal__entry-picture"
                    style={{
                        backgroundSize: "cover",
                        backgroundImage: `url(${ url })`
                    }}
                >
                </div>)
            }

            <div className="journal__entry-body">
                <p className="journal__entry-title">
                    { title }
                </p>
                <p className="journal__entry-content">
                    { body }
                </p>
            </div>

            <div className="journal__entry-date-box">
                <span>{ noteDate.format('dddd') }</span>
                <h4>{ noteDate.format('Do') }</h4>
            </div>

        </div>
    )
}
