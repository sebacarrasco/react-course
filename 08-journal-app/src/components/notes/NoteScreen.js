import React, { useEffect, useRef } from 'react';
import { NotesAppBar } from './NotesAppBar';
import { useDispatch, useSelector } from 'react-redux';
import { useForm } from '../../hooks/useForm';
import { activeNote, startDeleting } from '../../actions/notes';

export const NoteScreen = () => {

    const { active:note } = useSelector(state => state.notes);
    const dispatch = useDispatch();

    // Así no estaremos modificando la nota active, es como que creamos una copia
    const [ formValues, handleInputChange, reset ] = useForm(note);
    const { body, title } = formValues;

    // Recordar:
    // useRef es básicamente un setState pero que solo retorna el state
    // y se puede modificar directamente. Además, al modificarlo no se
    // vuelve a renderear.
    // Y por este ejercicio, al parecer no cambia su valor si el componente se
    // vuelve a renderear (aunque cambie el state de note.id, no se actualiza activeId)
    const activeId = useRef(note.id);

    // Este useEffect es para que se actualice el formulario al cambiar de nota
    // Por alguna razón al actualizar note, no se vuelve a llamr el useForm automáticamente
    useEffect(() => {
        if (note.id !== activeId.current)
        {
            reset(note);
            activeId.current = note.id
        }
    }, [note, reset])

    // Yo lo hubiera hecho así, pero por alguna razón se ejecuta el reset dos veces
    // al clickear una nota, incluso si es la nota que ya está activa.
    // Con la forma del profe se llama a reset solo una vez y de hecho cuando se clickea
    // la nota activa no se resetea c:
    // useEffect(() => {
    //     reset(note);
    // }, [note, reset])

    // Para actualizar la activeNote en la store: (ojo que no en la bdd)
    useEffect(() => {
        dispatch(activeNote(formValues.id, {...formValues}));
    }, [formValues, dispatch]);

    const handleDelete = () => {
        dispatch(startDeleting(note.id));
    }

    return (
        <div className="notes__main-content">
            <NotesAppBar />

            <div className="notes__content">

                <input
                    type="text"
                    placeholder="Some awesome title"
                    className="notes__title-input"
                    autoComplete="off"
                    name="title"
                    value={ title }
                    onChange={ handleInputChange }
                />

                <textarea
                    placeholder="What happened today"
                    className="notes__textarea"
                    name="body"
                    value={ body }
                    onChange={ handleInputChange }
                ></textarea>

                {
                    note.url &&
                    (<div className="notes__image">
                        <img
                            src={ note.url }
                            alt="some image"
                        />
                    </div>)
                }

            </div>


            <button
                className="btn btn-danger"
                onClick={ handleDelete }
            >
                Delete
            </button>
        </div>
    )
}
