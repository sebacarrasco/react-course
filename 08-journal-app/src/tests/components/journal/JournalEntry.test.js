import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { JournalEntry } from '../../../components/journal/JournalEntry';
import { activeNote } from '../../../actions/notes';

// Config de la store
const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {}
let store = mockStore(initState);
store.dispatch = jest.fn();

const note = {
    id: 20,
    date: 0,
    title: 'A title',
    body: 'The body of a note',
    url: 'https://someplace.com/pic.jpg'
}

describe('Tests for <JournalEntry />', () => {

    const wrapper = mount(
        <Provider store={ store }>
            <JournalEntry { ...note } />
        </Provider>
    );
   
    test('should match the snapshot', () => {
        
        expect(wrapper).toMatchSnapshot();

    });

    test('should active the note', () => {
       
        wrapper.find(".journal__entry").prop("onClick")();
        expect(store.dispatch).toHaveBeenCalledWith(activeNote(note.id, { ...note }));

    });
    
    

});
