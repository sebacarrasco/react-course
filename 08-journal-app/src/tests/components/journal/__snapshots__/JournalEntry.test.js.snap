// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for <JournalEntry /> should match the snapshot 1`] = `
<div
  className="journal__entry pointer animate__animated animate__fadeIn animate__faster"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url(https://someplace.com/pic.jpg)",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      A title
    </p>
    <p
      className="journal__entry-content"
    >
      The body of a note
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      Wednesday
    </span>
    <h4>
      31st
    </h4>
  </div>
</div>
`;
